src/components/SpreadsheetUploader/index.js

```diff
@@ -0,0 +1,189 @@
+import { Button, Card, Typography } from '@material-ui/core';
+import GordonDialogBox from 'components/GordonDialogBox';
+import GordonSnackbar from 'components/Snackbar';
+import { useEffect, useState } from 'react';
+import Dropzone from 'react-dropzone';
+import { Link } from 'react-router-dom';
+import { gordonColors } from 'theme';
+import XLSX from 'xlsx';
+import SpreadsheetSVG from './spreadsheet.svg';
+import styles from './SpreadsheetUploader.module.css';
+
+const SpreadsheetUploader = ({
+  open,
+  setOpen,
+  onSubmitData,
+  title,
+  maxColumns,
+  requiredColumns = [],
+  otherColumns = [],
+  buttonName,
+  template,
+}) => {
+  const [data, setData] = useState();
+  const [error, setError] = useState();
+
+  const acceptedTypes = ['application/vnd.ms-excel', 'text/csv'];
+
+  const onDropAccepted = (fileList) => {
+    let file = fileList[0];
+    if (!acceptedTypes.includes(file.type)) {
+      setError('The file is not one of the supported file types.');
+      return;
+    }
+    let reader = new FileReader();
+    reader.onload = function (event) {
+      let data = event.target.result;
+      let workbook = XLSX.read(data, {
+        type: 'binary',
+      });
+      let sheet = workbook.Sheets[workbook.SheetNames[0]];
+
+      let uploadedData = XLSX.utils.sheet_to_row_object_array(sheet);
+
+      /* For each row:
+          If the column count is less than required or greater than max
+          OR if there are some required columns not in the row
+          OR if there are some columns in the row that are not allowed in the object
+      */
+      if (
+        uploadedData.some((row) => {
+          let keys = Object.keys(row);
+          return (
+            keys.length > maxColumns ||
+            keys.length < requiredColumns.length ||
+            requiredColumns.some((column) => !keys.includes(column)) ||
+            keys.some((key) => !requiredColumns.includes(key) && !otherColumns.includes(key))
+          );
+        })
+      ) {
+        setError('Please make sure your file matches the template file.');
+        return;
+      }
```

Comment:
It took me a minute to understand that `keys` here represents the names of the columns of the given row. Could you use a more descriptive name, possibly `columns` or `columnNames`?

Reply:
Implementing changes:
```suggestion
      uploadedData.forEach((row) => {

        let columnNames = Object.keys(row);

        if (error) return;

        if (columnNames.length > maxColumns)
          setError("File has too many columns");
        if (columnNames.length < requiredColumns.length)
          setError("File has too few columns");
        if (requiredColumns.some((column) => !columnNames.includes(column)))
          setError("File is missing required columns");
        if (columnNames.some((columnName) => !requiredColumns.includes(columnName) && !otherColumns.includes(columnName)))
          setError("File contains extra columns");
      });
      
      if (error) return;
```

Reply:
That looks good! My only concern is that `error` is not updated synchronously by `setError`, so you may get some unexpected behavior here. There are two simple ways to resolve that:
1. Introduce a local `error` variable and then call `setError` outside the loop before returning if the local `error` is not null
2. Still use `Array.some` instead of `Array.forEach`, but still `setError` inside the loop. That way, you don't have to evaluate the whole dataset when you're only looking for the first error that exists.

Personally, I think that option #2 results in cleaner flow overall:
```suggestion
      const hasError = uploadedData.some((row) => {
            let columnNames = Object.keys(row);
	
	        if (columnNames.length > maxColumns) {
	          setError("File has too many columns");
	        } else if (columnNames.length < requiredColumns.length) {
	          setError("File has too few columns");
	        } else if (requiredColumns.some((column) => !columnNames.includes(column))) {
	          setError("File is missing required columns");
	        } else if (columnNames.some((columnName) => !requiredColumns.includes(columnName) && !otherColumns.includes(columnName))) {
	          setError("File contains extra columns");
	        } else {
	          return false;
	        }
	        
	        return true;
	      });
	      
	      if (hasError) return;
```